src/datachain/lib/video.py

```diff
@@ -0,0 +1,275 @@
+    if end_frame is not None:
+        if end_frame < 0:
+            raise ValueError("end_frame must be a non-negative integer.")
+        if start_frame > end_frame:
+            raise ValueError("start_frame must be less than or equal to end_frame.")
+    if step < 1:
+        raise ValueError("step must be a positive integer.")
+
+    # Compute the frame shift to determine the number of frames to skip,
+    # considering the start frame and step size
+    frame_shift = start_frame % step
+
+    # Iterate over video frames and yield only those within the specified range and step
+    for frame, img in enumerate(iio.imiter(file.stream(), plugin="pyav")):
+        if frame < start_frame:
+            continue
+        if (frame - frame_shift) % step != 0:
+            continue
+        if end_frame is not None and frame > end_frame:
+            break
+        yield img
+
+
+def video_frames(
+    file: "VideoFile",
+    start_frame: int = 0,
+    end_frame: Optional[int] = None,
+    step: int = 1,
+    format: str = "jpeg",
+) -> "Iterator[bytes]":
+    """
+    Reads video frames from a file and returns as bytes.
+
+    Args:
+        file (VideoFile): VideoFile object.
+        start_frame (int): Frame number to start reading from (default: 0).
+        end_frame (int): Frame number to stop reading at (default: None).
+        step (int): Step size for reading frames (default: 1).
+        format (str): Image format (default: 'jpeg').
+
+    Returns:
+        Iterator[bytes]: Iterator of video frames.
+    """
+    for img in video_frames_np(file, start_frame, end_frame, step):
+        yield iio.imwrite("<bytes>", img, extension=f".{format}")
+
+
+def save_video_frames(
+    file: "VideoFile",
+    output_dir: Union[str, pathlib.Path],
+    start_frame: int = 0,
+    end_frame: Optional[int] = None,
+    step: int = 1,
+    format: str = "jpeg",
+) -> "Iterator[str]":
+    """
+    Saves video frames as image files.
+
+    Args:
+        file (VideoFile): VideoFile object.
+        output_dir (Union[str, pathlib.Path]): Output directory path.
+        start_frame (int): Frame number to start reading from (default: 0).
+        end_frame (int): Frame number to stop reading at (default: None).
+        step (int): Step size for reading frames (default: 1).
+        format (str): Image format (default: 'jpeg').
+
+    Returns:
+        Iterator[str]: List of output file paths.
+    """
+    file_stem = file.get_file_stem()
+
+    for i, img in enumerate(video_frames_np(file, start_frame, end_frame, step)):
+        frame = start_frame + i * step
+        output_file = os.path.join(output_dir, f"{file_stem}_{frame:06d}.{format}")
+        iio.imwrite(output_file, img, extension=f".{format}")
+        yield output_file
+
+
+def save_video_clip(
+    file: "VideoFile",
+    start_time: float,
+    end_time: float,
+    output_file: Union[str, pathlib.Path],
+    codec: str = "libx264",
+    audio_codec: str = "aac",
+) -> None:
```

Comment:
It would be great to generalize the single and plural methods. We just need to come up with output format like `output="{name}{:06d}.{ext}")` and provide a string in case of a single file.

Also, this method will require generalization for writing to cloud like `output={source}/tmp/{name}{:06d}.{ext}`

Reply:
Agree, not done yet 🤔